core/src/main/java/io/kestra/core/runners/RunContextFactory.java

```diff
@@ -1,46 +1,154 @@
 package io.kestra.core.runners;
 
 import com.google.common.annotations.VisibleForTesting;
-import com.google.common.collect.ImmutableMap;
-import io.micronaut.context.ApplicationContext;
+import io.kestra.core.metrics.MetricRegistry;
 import io.kestra.core.models.executions.Execution;
 import io.kestra.core.models.executions.TaskRun;
 import io.kestra.core.models.flows.Flow;
 import io.kestra.core.models.tasks.Task;
 import io.kestra.core.models.triggers.AbstractTrigger;
-
-import java.util.Map;
+import io.kestra.core.plugins.PluginConfigurations;
+import io.kestra.core.storages.InternalStorage;
+import io.kestra.core.storages.StorageContext;
+import io.kestra.core.storages.StorageInterface;
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.context.annotation.Value;
 import jakarta.inject.Inject;
 import jakarta.inject.Singleton;
 
+import java.net.URI;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Optional;
+
 @Singleton
 public class RunContextFactory {
     @Inject
-    private ApplicationContext applicationContext;
+    protected ApplicationContext applicationContext;
+
+    @Inject
+    protected PluginConfigurations pluginConfigurations;
 
-    public RunContext of(Flow flow, Execution execution) {
-        return new RunContext(applicationContext, flow, execution);
+    @Inject
+    protected VariableRenderer variableRenderer;
+
+    @Inject
+    protected StorageInterface storageInterface;
+
+    @Inject
+    protected MetricRegistry metricRegistry;
+
+    @Inject
+    protected RunContextCache runContextCache;
+
+    @Value("${kestra.encryption.secret-key}")
+    protected Optional<String> secretKey;
+
+    @Value("${kestra.tasks.tmp-dir.path}")
+    protected Optional<String> tempBasedPath;
+
+    @Inject
+    private RunContextLoggerFactory runContextLoggerFactory;
+
+    // hacky
+    public RunContextInitializer initializer() {
+        return applicationContext.getBean(RunContextInitializer.class);
+    }
+
+    public MutableRunContext of(Flow flow, Execution execution) {
```

Comment:
Better to expose the interface here.
If you don't want to cast too often, a handy pattern is to provide a method `wrap(Class)` in the RunContext that will do the cast for you.